src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,712 @@
+#include <gtest/gtest.h>
+
+#include <QFile>
+#include <QFileInfo>
+#include <QTemporaryDir>
+#include <memory>
+
+#include "library/library_prefs.h"
+#include "sources/metadatasource.h"
+#include "test/librarytest.h"
+#include "track/track.h"
+#include "util/fileinfo.h"
+
+namespace {
+
+const auto kTestDir = QDir(QDir::current().absoluteFilePath("src/test/id3-test-data"));
+
+const QString kTestFileNameWithMetadata = QStringLiteral("cover-test-jpg.mp3");
+
+const QString kTestFileNameWithoutMetadata = QStringLiteral("empty.mp3");
+
+enum class FileType {
+    WithMetadata,
+    WithoutMetadata,
+};
+
+void sleepAfterFileLastModifiedUpdated() {
+    // Ensure that all subsequent modification time stamps will be
+    // strictly greater, i.e. strictly monotonic sequence of time
+    // stamps. This grace period seems to be mandatory for the CI
+    // test runs and might need to be adjusted depending on the
+    // results for different platforms.
+    QThread::msleep(1);
+}
+
+/// A temporary file system with a single audio file.
+class TempFileSystem {
+  public:
+    explicit TempFileSystem(const QString& fileName)
+            : m_fileInfo(m_tempDir.filePath(fileName)) {
+        createFile();
+    }
+
+    const mixxx::FileInfo& fileInfo() const {
+        return m_fileInfo;
+    }
+
+    QDateTime fileLastModified() const {
+        return m_fileInfo.toQFile().fileTime(QFileDevice::FileModificationTime);
+    }
+
+    void updateFileLastModified() const {
+        auto file = m_fileInfo.toQFile();
+        const auto oldLastModified = fileLastModified();
+        ASSERT_TRUE(oldLastModified.isValid());
+        // Rely on the OS to generate the time stamps instead of using
+        // QFileDevice::setFileTime() with a time stamp generated by
+        // QDateTime. The latter has proven to produce non-monotonic
+        // time stamps that cause spurious test failures.
+        ASSERT_TRUE(file.open(QIODevice::ReadWrite | QIODevice::ExistingOnly | QIODevice::Append));
+        const auto fileSize = file.size();
+        // Grow and shrink the file to bump the modification time stamp
+        ASSERT_TRUE(file.resize(fileSize + 1));
+        ASSERT_TRUE(file.flush());
+        ASSERT_TRUE(file.resize(fileSize));
+        file.close();
+        const auto newLastModified = fileLastModified();
+        ASSERT_TRUE(newLastModified.isValid());
+        ASSERT_GT(newLastModified, oldLastModified);
+        sleepAfterFileLastModifiedUpdated();
```

Comment:
Calling sleep in a function called updateFileLastModified() is supprising. 
I think this should be moved to the outer scope to understand the issue with > and =. 
Or can we dispose this?

Reply:
This also applies to all other invocations of sleepAfterFileLastModifiedUpdated();
If this is essential to test is shall be called in the test to understand the logic in place.

Reply:
The test cases should focus on the business logic and not be bothered with technical workarounds. That's one purpose of those utility classes.

Reply:
https://github.com/mixxxdj/mixxx/pull/4628#discussion_r788003336

Reply:
I disagree here. The user should know in the test what he has to expect from the timing. This is essential, because the test may fail because of timing issues.